c10/util/intrusive_ptr.h

```diff
@@ -31,43 +26,25 @@ namespace raw {
 // was dynamically allocated (e.g., via make_intrusive).
 //
 // Thus, whenever you transmute a T* into a intrusive_ptr<T>, we check
-// and make sure that the refcount isn't zero (or, a more subtle
-// test for weak_intrusive_ptr<T>, for which the refcount may validly
-// be zero, but the weak refcount better not be zero), because that
+// and make sure that the refcount isn't zero, because that
 // tells us if the object was allocated by us.  If it wasn't, no
 // intrusive_ptr for you!
 
 class C10_API intrusive_ptr_target {
-  // Note [Weak references for intrusive refcounting]
-  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-  // Here's the scheme:
-  //
-  //  - refcount == number of strong references to the object
-  //    weakcount == number of weak references to the object,
-  //      plus one more if refcount > 0
-  //    An invariant: refcount > 0  =>  weakcount > 0
-  //
-  //  - THStorage stays live as long as there are any strong
-  //    or weak pointers to it (weakcount > 0, since strong
-  //    references count as a +1 to weakcount)
-  //
-  //  - finalizers are called and data_ptr is deallocated when refcount == 0
-  //
-  //  - Once refcount == 0, it can never again be > 0 (the transition
-  //    from > 0 to == 0 is monotonic)
-  //
-  //  - When you access THStorage via a weak pointer, you must
-  //    atomically increment the use count, if it is greater than 0.
-  //    If it is not, you must report that the storage is dead.
-  //
   mutable std::atomic<size_t> refcount_;
-  mutable std::atomic<size_t> weakcount_;
+  static std::atomic<int64_t> unique_id_counter_;
+  int64_t unique_id_;
 
   template <typename T, typename NullType>
   friend class intrusive_ptr;
-  template <typename T, typename NullType>
-  friend class weak_intrusive_ptr;
-  friend inline void raw::weak_intrusive_ptr::incref(intrusive_ptr_target* self);
+
+ public:
+  int64_t unique_id() {
```

Comment:
unique_id() should be "const" as a public interface. 

unique_id_ should be marked as "mutable" since it is internally a caching mechanism.